### PR TITLE
Enable abs of a param real/imag to produce a param real

### LINKED ISSUE #24125: feature request: abs for param complex
### Summary of Problem
`abs` on a param complex number does not produce a param, but I would expect it to.

In order to implement this, we will a need param-time `sqrt` for `real` numbers. That should also be available to users.

### Steps to Reproduce

**Source Code:**

```chapel
param c:complex(128) = 1.0i;    
param x = abs(c); // error: Initializing parameter 'x' to value not known at compile time
```

**Associated Future Test(s):**

[`test/library/standard/AutoMath/absparamscomplex.chpl`](
           https://github.com/chapel-lang/chapel/blob/main/test/library/standard/AutoMath/absparamscomplex.chpl
          ) #24127

### PR DESCRIPTION
**Pull request by mppf**
For issue #24119.
Resolves #24125.

This PR adds overloads to AutoMath to implement `param` versions of:
 * `abs` for all numeric types
 * `sqrt` for floating-point types.
 
Additionally, while testing, I realized that `complex.re` and `complex.im` did not have `param` returning versions, so this PR also adds such param-returning versions to ChapelBase.

To implement these, I:
 * added `PRIM_SQRT` and `PRIM_ABS`. For now, these are just used for computing these functions on `param`s and they will cause the compiler to error out one makes it to code-generation time.
 * implemented folding for `PRIM_ABS`, `PRIM_SQRT`, `PRIM_GET_REAL`, and `PRIM_GET_IMAG` in num.cpp
 * enabled `postFoldPrimop` to fold `PRIM_ABS`, `PRIM_SQRT`, `PRIM_GET_REAL`, and `PRIM_GET_IMAG`

This PR also updates the spec description of `complex` to make it clearer. While there, it removes `proc complex.bits` from the spec, since this method does not exist other than in the spec.

Reviewed by @DanilaFe - thanks!

- [x] full comm=none testing

Future Work:
 * #22809
   * in particular, resolve two confusing errors I posted as comments based on work on this PR
 * update the dyno resolver to assert `PRIM_ABS`, `PRIM_SQRT` are only used on `param` values
 * update the dyno resolver to implement `PRIM_GET_REAL` and `PRIM_GET_IMAG` on `param` values

### CODE REVIEW

**Comment by damianmoz** [comment]
Not urgent

File: modules/standard/AutoMath.chpl [hunk 1]
```diff
@@ -217,9 +222,20 @@ module AutoMath {
     return fabsf(x);
   }
 
+  /* Return the absolute value of a param real(32) as a param */
+  proc abs(param x: real(32)) param : real(32) {
+    return if x < 0 then -x else x;
+  }
```
Comment (DanilaFe) [inline modules/standard/AutoMath.chpl:228]: Floating points are weird, and this worries me a little. Consider the following Python:

```Python
>>> float("-0")
-0.0
>>> float("-0") < 0
False
>>> abs(float("-0"))
0.0
```

In other words, numerically this differs from the behavior of Python's `abs`:

```Python
>>> def myabs(x): return -x if x < 0 else x
...
>>> myabs(float("-0"))
-0.0
>>> abs(float("-0"))
0.0
```

Comment (mppf) [reply modules/standard/AutoMath.chpl:228]: I can wire this up to call an `abs` primitive in the compiler to run `fabs`. Do you feel that would address the issue?

Comment (DanilaFe) [reply modules/standard/AutoMath.chpl:228]: Yes, I do -- `fabs` seems ideal.

**Comment by mppf** [comment]
@DanilaFe - OK, I've added the primitive. From this point it would be trivial to also add param `abs` support for `complex` as well as param `sqrt` support. Do you think I should proceed with those in this PR?

**Comment by DanilaFe** [comment]
Sorry for the delayed response. I think you shouldn't worry about sqrt and complex here -- it can be left for future work (a "good first issue" perhaps?).

File: frontend/lib/resolution/prims.cpp [hunk 2]
```diff
@@ -740,11 +740,29 @@ primComplexGetComponent(Context* context, const CallInfo& ci) {
...
+  QualifiedType actualType = ci.actual(0).type();
+  if (auto comp = actualType.type()->toComplexType()) {
+    int w = comp->componentBitwidth();
+    ret = QualifiedType(QualifiedType::CONST_VAR, RealType::get(context, w));
+  } else if (auto img = actualType.type()->toImagType()) {
+    int w = img->bitwidth();
+    ret = QualifiedType(QualifiedType::CONST_VAR, RealType::get(context, w));
+  } else {
+    // otherwise just use the original type
+    ret = QualifiedType(QualifiedType::CONST_VAR, actualType.type());
```
Comment (DanilaFe) [inline frontend/lib/resolution/prims.cpp:765]: Note that the production compiler treats ABS and SQRT as param-only primitives. That leads me to think that Dyno should also implement the param logic, and not support the non-param case.

Comment (mppf) [reply frontend/lib/resolution/prims.cpp]: OK, I added some comments along these lines, and will note it in Future Work.

**Comment by bradcray** [comment]
Nice, thanks!

**Pull request merged**